File: csharp/queen-attack/QueenAttack.cs
using System;

public class Queen
{
    public Queen(int row, int column)
    {
        Row = row;
        Column = column;
    }

    public int Row { get; }
    public int Column { get; }
}

public static class QueenAttack
{
    private const int BoardSize = 8;
    
    public static bool CanAttack(Queen white, Queen black) =>
        IsSameRow(white, black) ||
        IsSameColumn(white, black) ||
        IsSameDiagonal(white, black);

    public static Queen Create(int row, int column)
    {
        if (!IsInBounds(row))
        {
            throw new ArgumentOutOfRangeException(nameof(row), row, "Row is out of bounds");
        }
        
        if (!IsInBounds(column))
        {
            throw new ArgumentOutOfRangeException(nameof(column), column, "Column is out of bounds");
        }
        
        return new Queen(row, column);
    }

    private static bool IsInBounds(int n) => 
        n >= 0 && n < BoardSize;
    
    private static bool IsSameRow(Queen white, Queen black) => 
        white.Row == black.Row;
    
    private static bool IsSameColumn(Queen white, Queen black) => 
        white.Column == black.Column;
    
    private static bool IsSameDiagonal(Queen white, Queen black) => 
        Math.Abs(white.Row - black.Row) == Math.Abs(white.Column - black.Column);
}

File: c/queen-attack/queen_attack.c
#include "queen_attack.h"
#include <stdbool.h>
#include <stdlib.h>

const uint8_t board_size = 8;

static bool is_valid(position_t queen)
{
    return queen.row < board_size && queen.column < board_size;
}

static bool is_same_row(position_t queen_1, position_t queen_2)
{
    return queen_1.row == queen_2.row;
}

static bool is_same_column(position_t queen_1, position_t queen_2)
{
    return queen_1.column == queen_2.column;
}

static bool is_same_diagonal(position_t queen_1, position_t queen_2)
{
    return abs(queen_2.row - queen_1.row) ==
           abs(queen_2.column - queen_1.column);
}

attack_status_t can_attack(position_t queen_1, position_t queen_2)
{
    if (!is_valid(queen_1) ||
        !is_valid(queen_2) ||
        (is_same_row(queen_1, queen_2) && is_same_column(queen_1, queen_2)))
    {
        return INVALID_POSITION;
    }

    if (is_same_row(queen_1, queen_2) ||
        is_same_column(queen_1, queen_2) ||
        is_same_diagonal(queen_1, queen_2))
    {
        return CAN_ATTACK;
    }

    return CAN_NOT_ATTACK;
}
